### PR TITLE
Numpy Support for access to mediagrains data

### PR DESCRIPTION
**Pull request by jamesba**
This PR implements numpy support for mediagrains grains in a variety of video formats, but only in py3.6+.

Job: https://www.pivotaltracker.com/story/show/168537286

### CODE REVIEW

**Comment by jamesba** [comment]
Have put in first pull request, which covers the numpy access methods. The grain format conversion code will be a separate PR building on this.

File: mediagrains_py36/numpy/__init__.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+        return np.dtype(np.int32)
+
+    raise NotImplementedError("Cog Frame Format not amongst those supported for numpy array interpretation")
+
+
+def _component_arrays_for_data_and_type(data: np.ndarray, fmt: CogFrameFormat, components: bytesgrain.VIDEOGRAIN.COMPONENT_LIST):
+    if not COG_FRAME_IS_PACKED(fmt) and not COG_FRAME_IS_COMPRESSED(fmt):
+        arrays = []
+        for component in components:
+            component_data = data[component.offset//data.itemsize:(component.offset + component.length)//data.itemsize]
```
Comment (simonrankine) [inline mediagrains_py36/numpy/__init__.py]: Huh, well the floor quotient operator is a bit of syntax I never knew...

File: mediagrains_py36/numpy/__init__.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        return np.dtype(np.int16)
+    elif fmt == CogFrameFormat.v210:
+        return np.dtype(np.int32)
+
+    raise NotImplementedError("Cog Frame Format not amongst those supported for numpy array interpretation")
+
+
+def _component_arrays_for_data_and_type(data: np.ndarray, fmt: CogFrameFormat, components: bytesgrain.VIDEOGRAIN.COMPONENT_LIST):
+    if not COG_FRAME_IS_PACKED(fmt) and not COG_FRAME_IS_COMPRESSED(fmt):
+        arrays = []
```
Comment (simonrankine) [inline mediagrains_py36/numpy/__init__.py]: Arrays of what?

File: mediagrains_py36/numpy/__init__.py [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+                       shape=(components[0].height, components[0].width),
+                       strides=(components[0].stride, data.itemsize*4)).transpose(),
+            as_strided(data[2:],
+                       shape=(components[0].height, components[0].width),
+                       strides=(components[0].stride, data.itemsize*4)).transpose(),
+            as_strided(data[3:],
+                       shape=(components[0].height, components[0].width),
+                       strides=(components[0].stride, data.itemsize*4)).transpose()]
+    elif fmt == CogFrameFormat.v210:
+        # v210 is barely supported. Convert it to something else to actually use it!
```
Comment (simonrankine) [inline mediagrains_py36/numpy/__init__.py]: As you just return an empty array here, maybe better to raise a NotImplemented exception instead? This seems like un-expected behaviour.

File: mediagrains_py36/numpy/__init__.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+
+def _component_arrays_for_data_and_type(data: np.ndarray, fmt: CogFrameFormat, components: bytesgrain.VIDEOGRAIN.COMPONENT_LIST):
+    if not COG_FRAME_IS_PACKED(fmt) and not COG_FRAME_IS_COMPRESSED(fmt):
+        arrays = []
+        for component in components:
+            component_data = data[component.offset//data.itemsize:(component.offset + component.length)//data.itemsize]
+            component_data = as_strided(component_data, shape=(component.height, component.width), strides=(component.stride, component_data.itemsize))
+            arrays.append(component_data.transpose())
+        return arrays
+    elif fmt in [CogFrameFormat.UYVY, CogFrameFormat.v216]:
```
Comment (simonrankine) [inline mediagrains_py36/numpy/__init__.py]: The block that follows this feels a bit messy, could it be re-factored into a method that accepts a block of data, the width and itemsize for it?

File: mediagrains/cogenums.py [hunk 5]
```diff
@@ -117,10 +132,30 @@ def COG_FRAME_IS_COMPRESSED(fmt):
     return ((fmt >> 9) & 0x1) != 0
 
 
+def COG_FRAME_IS_ALPHA(fmt):
+    return ((fmt >> 7) & 0x1) != 0
+
+
+def COG_FRAME_IS_RGB(fmt):
```
Comment (philipnbbc) [inline mediagrains/cogenums.py]: Change this to `COG_FRAME_IS_PLANAR_RGB` given that it only applies to planar?

File: mediagrains/cogenums.py [hunk 6]
```diff
@@ -25,7 +25,17 @@
 
 from enum import IntEnum
 
-__all__ = ['CogFrameFormat', 'CogFrameLayout', 'CogAudioFormat', 'COG_FRAME_IS_PACKED', 'COG_FRAME_IS_COMPRESSED', 'COG_FRAME_FORMAT_BYTES_PER_VALUE']
+__all__ = [
+    'CogFrameFormat',
+    'CogFrameLayout',
+    'CogAudioFormat',
+    'COG_FRAME_IS_PACKED',
```
Comment (philipnbbc) [inline mediagrains/cogenums.py]: Add a `COG_FRAME_IS_PLANAR` to simplify `not COG_FRAME_IS_PACKED(fmt) and not COG_FRAME_IS_COMPRESSED(fmt)`?

File: mediagrains_py36/numpy/__init__.py [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+    if not COG_FRAME_IS_PACKED(fmt) and not COG_FRAME_IS_COMPRESSED(fmt):
+        if COG_FRAME_IS_RGB(fmt):
+            return ComponentDataList.ComponentOrder.RGB
+        else:
+            return ComponentDataList.ComponentOrder.YUV
+    elif fmt in [CogFrameFormat.UYVY, CogFrameFormat.YUYV, CogFrameFormat.v216]:
+        return ComponentDataList.ComponentOrder.YUV
+    elif fmt in [CogFrameFormat.RGB, CogFrameFormat.RGBA, CogFrameFormat.RGBx, CogFrameFormat.ARGB, CogFrameFormat.xRGB]:
+        return ComponentDataList.ComponentOrder.RGB
+    elif fmt in [CogFrameFormat.BGRA, CogFrameFormat.BGRx]:
```
Comment (philipnbbc) [inline mediagrains_py36/numpy/__init__.py]: Also `xBGR` and `ABGR`?

File: mediagrains_py36/numpy/__init__.py [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+def _dtype_from_cogframeformat(fmt: CogFrameFormat) -> np.dtype:
+    if not COG_FRAME_IS_PACKED(fmt) and not COG_FRAME_IS_COMPRESSED(fmt):
+        if COG_FRAME_FORMAT_BYTES_PER_VALUE(fmt) == 1:
+            return np.dtype(np.uint8)
+        elif COG_FRAME_FORMAT_BYTES_PER_VALUE(fmt) == 2:
+            return np.dtype(np.int16)
+        elif COG_FRAME_FORMAT_BYTES_PER_VALUE(fmt) == 4:
+            return np.dtype(np.int32)
+    elif fmt in [CogFrameFormat.UYVY,
+                 CogFrameFormat.YUYV,
```
Comment (philipnbbc) [inline mediagrains_py36/numpy/__init__.py]: `AYUV`

File: mediagrains_py36/numpy/__init__.py [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+            self.R = self[2]
+
+
+def _component_arrangement_from_format(fmt: CogFrameFormat):
+    if not COG_FRAME_IS_PACKED(fmt) and not COG_FRAME_IS_COMPRESSED(fmt):
+        if COG_FRAME_IS_RGB(fmt):
+            return ComponentDataList.ComponentOrder.RGB
+        else:
+            return ComponentDataList.ComponentOrder.YUV
+    elif fmt in [CogFrameFormat.UYVY, CogFrameFormat.YUYV, CogFrameFormat.v216]:
```
Comment (philipnbbc) [inline mediagrains_py36/numpy/__init__.py]: `AYUV`?

File: mediagrains/grain.py [hunk 10]
```diff
@@ -793,7 +796,9 @@ class COMPONENT(Mapping):
 length
     The total length of the data for this component in bytes
 """
-        def __init__(self, meta):
+        def __init__(self, parent, index, meta):
```
[the remark this answers is not in the document]
Comment (philipnbbc) [reply mediagrains/grain.py]: It's not immediately obvious why the COMPONENT needs the parent and list index. Is this in preparation for code to come or maybe for code consistency?

**Comment by philipnbbc** [comment]
Looks fine to me in general. Using bit 4 in the format word to indicate RGB/YUV for planar makes sense. I had a couple of inline comments, but nothing major.

File: mediagrains/grain.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (jamesba) [reply mediagrains/grain.py]: It's in preparation for code that was to come ... and then was removed.

[the remark this answers is not in the document]
Comment (jamesba) [reply mediagrains/grain.py]: I will remove this.

File: mediagrains_py36/numpy/__init__.py (hunk 1 above)
Comment (jamesba) [reply mediagrains_py36/numpy/__init__.py]: It was rarely used in Python 2, but becomes more important in Py3, since `int/int=float` but `int//int=int`.

**Pull request merged**